lib/elixir/lib/file.ex

```diff
@@ -310,24 +310,23 @@ defmodule File do
   end
 
   defp do_mkdir_p(path) do
-    if dir?(path) do
+    if path == "/" or path == "." do
       :ok
     else
       parent = Path.dirname(path)
 
-      if parent == path do
-        # Protect against infinite loop
-        {:error, :einval}
-      else
-        _ = do_mkdir_p(parent)
-
+      with {_, false} <- {:loop, parent == path},
```

Comment:
Please don't  use tagged tuples. If you need to match on the results, then it is best to match using nested cases.